### PR TITLE
DM-21838: Add some Table/FitsTable methods

### PR DESCRIPTION
**Pull request by timj**
Sufficient for creating FITS -TAB WCS.

### CODE REVIEW

File: python/astshim/fitsChanContinued.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+FitsChan.__contains__ = contains
+
+
+def getitem(self, name):
+    """Return the value if a keyword is specified, or the entire card if
+    an integer index is specified."""
+
+    # Save current card position
+    currentCard = self.getCard()
```
Comment (r-owen) [inline python/astshim/fitsChanContinued.py:182]: Please consider a try/finally block for the remaining code to make it more obvious (and more reliable in the face of future code changes) that the current card will always be restored:
```
try:
    # do stuff
finally:
    self.setCard(currentCard)
````
I suggest this same pattern for all other methods that preserve (or try to preserve) the current card.

File: python/astshim/fitsChanContinued.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+                raise RuntimeError(f"Type, {ctype} of FITS card '{name}' not supported")
+
+            # Increment the card number to continue search
+            self.setCard(thiscard + 1)
+
+        # Reinstate the original card position
+        self.setCard(currentCard)
+
+        # We may have multiple values. For consistency with pyast we return
+        # a single item for a single match, else a tuple
```
Comment (r-owen) [inline python/astshim/fitsChanContinued.py]: I would be *much* happier if the return value was always a tuple so the user doesn't have to check. In this case I think it's worth breaking the precent set by pyast. This "sometimes a tuple sometimes a scalar" business has bitten us badly with PropertyList/PropertySet and I do not want to repeat that experience.

File: python/astshim/fitsChanContinued.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        raise StopIteration
+    return card.value
+
+
+FitsChan.__next__ = next
+
+
+def contains(self, name):
+    """Returns True if either the supplied name is present in the FitsChan
+    or the supplied integer is acceptable to the FitsChan."""
```
Comment (r-owen) [inline python/astshim/fitsChanContinued.py]: Nit-pick but our coding standard  is to have a newline before the final `"""` in a doc string: https://developer.lsst.io/python/numpydoc.html#docstrings-should-begin-with-and-terminate-with-on-its-own-line

File: python/astshim/fitsChanContinued.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+
+FitsChan.__getitem__ = getitem
+
+
+def setitem(self, name, value):
+    """name can be integer index or keyword.
+
+    If an integer, the value is deemed to be a full card to replace the
+    existing one. If the value is None or empty string the card is deleted.
```
Comment (r-owen) [inline python/astshim/fitsChanContinued.py]: This does not seem to match the code below: if I am reading the code correctly `None` makes the value undefined, which seems like a much better choice than deleting the card.

Also an empty string seems to set the card value to an empty string. If that is legal FITS then that also seems like the best choice. If it is not legal FITS then I guess go with undefined?

I would much rather use a different call to delete items.

**Review by r-owen — CHANGES_REQUESTED** [review]
I really appreciate the new dict access for FitsChan, though I am requesting that you always return a tuple of values when specifying a card name.

Also please add more tests for Table and FitsTable if possible. I'd like to see all methods explicitly exercised, preferably in test_table and test_fitsTable files so I know where to look for them.

Also please explicitly testing setting a card to an empty string.

**Comment by timj** [comment]
I don't have explicit tests for Table and FitsTable because I haven't implemented enough of the API to be able to write any tests. This ticket is mostly implementing the minimum required to support -TAB WCS. Adding all of Table and FitsTable will be a lot more work and I'm not sure I can justify doing that.

**Comment by r-owen** [comment]
That is fair alas. I'd still like the dict get name to return a tuple (or if you object, let's discuss this more widely), the requested additional tests of the dict interface and fixing the doc string for the method that sets values that I noted.

File: python/astshim/fitsChanContinued.py (hunk 4 above)
Comment (timj) [reply python/astshim/fitsChanContinued.py]: The distinction is that if you set a card by position (ie with an integer index) to `None` it is deleted. If you set it by name to `None` you get an undefined value. Assigning `None` to a position is generally ambiguous since there is no undef equivalent (remember that assigning by index means you are assigning an entire card, not a value for the header item). A completely blank header would be an empty string.  I think the current logic is reasonable in this case.

Comment (r-owen) [reply python/astshim/fitsChanContinued.py]: The doc string says "If the value is None or empty string the card is deleted". And you just said in your comment "If you set it by name to `None` you get an undefined value". The latter (keeping the card but making the value undefined) seems like much more desirable behavior.

Comment (timj) [reply python/astshim/fitsChanContinued.py]: The first paragraph was solely talking about the integer index case. The next paragraph talks about the header name case. Regardless, I have removed the delete and added `__delitem__`.

**Comment by timj** [comment]
@r-owen I've made all the requested changes if you would like to take another look.

File: python/astshim/fitsChanContinued.py [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+def length(self):
+    return self.nCard
+
+
+FitsChan.__len__ = length
+
+
+def iter(self):
+    """The FitsChan itself is the iterator.
+    """
```
Comment (r-owen) [inline python/astshim/fitsChanContinued.py:96]: But what is the result of the iteration (as per the comment for `next` below)?

File: python/astshim/fitsChanContinued.py [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+    self.clearCard()
+    return self
+
+
+FitsChan.__iter__ = iter
+
+
+def next(self):
+    """Return the entire header card until we run out of cards.
```
Comment (r-owen) [inline python/astshim/fitsChanContinued.py]: Please clarify what this returns.

I would thing the best thing to return is an object from which one can get the key, the value and the 80 character formatted string. But reading the code it appears that it returns just one of those (based on to_string I am *guessing* it is the 80 character string, despite returning `card.value` which I would naively expect to be the value).

File: python/astshim/fitsChanContinued.py [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+            return True
+    elif isinstance(name, str):
+        currentCard = self.getCard()
+        try:
+            self.clearCard()
+            result = self.findFits(name, False)
+        finally:
+            self.setCard(currentCard)
+        if result.found:
+            return True
```
Comment (r-owen) [inline python/astshim/fitsChanContinued.py:142]: Please raise an exception if `name` is neither int nor string. The current code silently does nothing.

File: python/astshim/fitsChanContinued.py [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+
+FitsChan.__getitem__ = getitem
+
+
+def setitem(self, name, value):
+    """name can be integer index or keyword.
+
+    If an integer, the value is deemed to be a full card to replace the
+    existing one. If the value is None or empty string a blank header
+    card is created at the location.
```
Comment (r-owen) [inline python/astshim/fitsChanContinued.py]: Sorry to keep returning to this, but I think this needs a bit more work, both in implementation and documentation.

Regarding value being None or the empty string:
* If name is an integer then the value being None or empty seems not useful. Perhaps raise an exception? What is the point to adding an empty card? (I think you were deleting the item before I complained the first time.)
* If name is a string then the value being None presumably means "value is undefined" (if so, please say so). Can a FITS card be string-valued and have a value that is the empty string? If not, then I suggest raising an exception.

Second: what really happens if name is a string, in the following cases:
* No such card exists: you create a new card?
* One matching card exists: you replace the value?
* Multiple matching cards exist: what??? I suggest raising an exception because "when in doubt refuse to guess".

Finally, documentation:
* Please have separate paragraphs for the "integer" and "string" case which clearly explain the above.
* Please document what happens if name is an integer and is out of range. My guess is it raises an exception, and that would be simplest, but which one?

File: python/astshim/fitsChanContinued.py [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+
+
+FitsChan.__setitem__ = setitem
+
+
+def delitem(self, name):
+    """Delete an item from the FitsChan either by index (0-based) or by name.
+
+    If a name is given all instances of the name will be deleted.
+    The current card position is not retained.
```
Comment (r-owen) [inline python/astshim/fitsChanContinued.py]: This is great. One minor request: please document what happens if name is an integer and is out of bounds. I hope it raises an exception, but which one?

Also please clarify what happens if name is a string and does not match any cards. I am guessing it silently does nothing, but please say so.

**Review by r-owen — CHANGES_REQUESTED** [review]
This looks a lot better but I still want some clarity in doc strings for the Python interface and would prefer slightly different behavior from setitem.

File: python/astshim/fitsChanContinued.py (hunk 6 above)
Comment (timj) [reply python/astshim/fitsChanContinued.py]: It returns the 80 character header card which is the fundamental unit of a FITS header.  This matches the standard convention of pyast.  Returning just the keys can't work because they can match multiple values and returning an integer index is superfluous. I'll see how it looks if I return a readonly dataclass FitsCard.

I would expect an `items()` method to return a tuple of keyword and value (which I can add).

File: python/astshim/fitsChanContinued.py (hunk 8 above)
Comment (timj) [reply python/astshim/fitsChanContinued.py]: Adding empty cards (80 characters of spaces) is a reasonable thing to do and are supported by the FITS standard. The question is whether None should imply empty car or raise an exception (an empty string is clearly an empty card).

It is clear that for named card a None means Undefined card (we use that all the time in LSST in PropertyList).  FITS empty string is well defined.

As you can see from the code if you have an existing card with multiple matches the later ones are deleted.

I think integer out of range leads to the card being added to the end (it's an append() operation).

File: python/astshim/fitsChanContinued.py (hunk 9 above)
Comment (timj) [reply python/astshim/fitsChanContinued.py]: `del` should probably raise an exception if the requested key is not found (as happens in a normal dict).

File: python/astshim/fitsChanContinued.py (hunk 6 above)
Comment (r-owen) [reply python/astshim/fitsChanContinued.py]: Thank you for the explanation. I agree the behavior is reasonable. Please document it in the doc string.

Adding an `items` method would be *really* helpful.

File: python/astshim/fitsChanContinued.py (hunk 8 above)
Comment (r-owen) [reply python/astshim/fitsChanContinued.py]: Thank you for the explanation.

Consider the following to make the integer and string cases behave more similarly:
- For an integer index None means make the value of the card undefined.
- For an integer index "" means make the value an empty string.

In both cases key is left unchanged, just as if `name` was a string.

I am uneasy about "arbitrary nonexistent index = append a card". What would you think about making "index = length" do that, but any other non-existent value raises a range error?

The string behavior seems reasonable. I'm not thrilled about deleting multiple cards, but can live with it.

Comment (timj) [reply python/astshim/fitsChanContinued.py]: I don't know what it means to make a card undefined. Integer index solely works with full FITS 80 character cards and not values of keywords. The only alternatives are therefore None to mean 80 spaces (just like an empty string) or None to raise an exception. It always replaces what was in that slot before.

File: python/astshim/fitsChanContinued.py (hunk 7 above)
Comment (timj) [reply python/astshim/fitsChanContinued.py:142]: I disagree. Contains is meant to be telling you if the thing is in the container. It should not raise if the thing is not in the container, it should return False, which it does do.

Comment (r-owen) [reply python/astshim/fitsChanContinued.py:142]: An interesting point. On the one hand for real dicts there is no such exception, but for real dicts the key can be any immutable type so that makes sense. From my perspective Python generally warns you when you try to use the wrong type so I think a TypeError would make a lot of sense here. But I can see your point and will not insist.

**Comment by r-owen** [comment]
Tim and I talked about the Python dict-like interface to FitsChan and concluded:
* Tim will add an `items()` method.
* An integer index is always used to get or set a formatted 80 character FITS card.
* A string index is always used to get or set a value (e.g. a float, int or whatever).
* Tim will update the doc strings for `next`, `getitem` and `setitem` to make this distinction clear.
* `getitem` with a string name will return the value for the first card, if there are multiple cards with that name. The only way to get all values is to use the `items()` method.

**Comment by timj** [comment]
@r-owen I believe I have made the agreed changes.

File: python/astshim/fitsChanContinued.py [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+    elif abs(index) > nCards:
+        # We allow index of one higher to indicate append
+        raise IndexError(f"Index {index} exceeds size of FitsChan ({nCards})")
+
+    # Convert to 1-based index
+    return index + 1
+
+
+def _get_current_card_value(self):
+    """Retrieve the value of the current card.
```
Comment (r-owen) [inline python/astshim/fitsChanContinued.py]: Please change "value" to "name and value" since it returns both.

File: python/astshim/fitsChanContinued.py (hunk 5 above)
Comment (r-owen) [reply python/astshim/fitsChanContinued.py:96]: Please document what the result of iteration actually is (e.g. if I do `for x in fitschan.iter()` what do I get)?

File: python/astshim/fitsChanContinued.py [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+
+    Parameters
+    ----------
+    name : `str` or `int`
+        If the FitsChan is being accessed by integer index the returned value
+        will be the corresponding 80-character card.  Index values are 0-based.
+        A negative index counts from the end of the FitsChan.
+        If the FitsChan is being accessed by string the returned value will
+        be the scalar value associated with the first card that matches the
+        supplied name.
```
Comment (r-owen) [inline python/astshim/fitsChanContinued.py:161]: Thank you. Crystal clear.

File: python/astshim/fitsChanContinued.py [hunk 12]
```diff
@@ -0,0 +1,378 @@
...
+FitsChan.__getitem__ = getitem
+
+
+def setitem(self, name, value):
+    """Set a new value.
+
+    Parameters
+    ----------
+    name : `str` or `int`
+        If the name is an integer index the returned value is the corresponding
```
Comment (r-owen) [inline python/astshim/fitsChanContinued.py]: "returned value"? I think you have a bit of a copy-paste issue here. Most of this text is fine but there are a few bits that need cleanup for this being getitem instead of setitem. In particular it still does not say (clearly enough) that when name is an integer then the value must be an 80 character formatted card.

Which raises a question I forgot to ask earlier: what happens if `name` is an integer and `value` is **not** a string, or is a string but is not 80 characters long? I hope you'd raise an exception.

File: python/astshim/fitsChanContinued.py [hunk 13]
```diff
@@ -0,0 +1,378 @@
...
+    """Iterate over each card, returning the keyword and value in a tuple.
+
+    Returns
+    -------
+    key : `str`
+        The key associated with the card.  Can be an empty string for some
+        comment styles.  The same key name can be returned multiple times
+        and be associated with different values.
+    value : `str`, `int`, `bool`, `float`
+        The value.
```
Comment (r-owen) [inline python/astshim/fitsChanContinued.py:375]: Or `None` for a key with an undefined value?!

**Review by r-owen — APPROVED** [review]
A few small requests for clarification and possibly one additional check/raise (setitem with an integer index and a value that is not an 80 character string)
